besu/src/main/java/org/hyperledger/besu/controller/TransitionBesuControllerBuilder.java

```diff
@@ -176,6 +189,50 @@ protected PluginServiceFactory createAdditionalPluginServices(
     return new NoopPluginServiceFactory();
   }
 
+  @Override
+  protected Synchronizer createSynchronizer(
+      final ProtocolSchedule protocolSchedule,
+      final WorldStateStorage worldStateStorage,
+      final ProtocolContext protocolContext,
+      final Optional<Pruner> maybePruner,
+      final EthContext ethContext,
+      final SyncState syncState,
+      final EthProtocolManager ethProtocolManager,
+      final PivotBlockSelector pivotBlockSelector) {
+
+    DefaultSynchronizer sync =
+        (DefaultSynchronizer)
+            super.createSynchronizer(
+                protocolSchedule,
+                worldStateStorage,
+                protocolContext,
+                maybePruner,
+                ethContext,
+                syncState,
+                ethProtocolManager,
+                pivotBlockSelector);
+    final GenesisConfigOptions maybeForTTD = configOptionsSupplier.get();
+    Optional<Difficulty> currentTotal =
+        protocolContext
+            .getBlockchain()
+            .getTotalDifficultyByHash(protocolContext.getBlockchain().getChainHeadHash());
+
+    PandaPrinter.init(
+        currentTotal,
+        Difficulty.of(configOptionsSupplier.get().getTerminalTotalDifficulty().get()));
+    if (maybeForTTD.getTerminalTotalDifficulty().isPresent()) {
+      LOG.info(
+          "TTD present, creating DefaultSynchronizer that stops propagating after finalization");
+      protocolContext.getConsensusContext(MergeContext.class).addNewForkchoiceMessageListener(sync);
+    }
+    sync.subscribeInSync(PandaPrinter.getInstance());
+    protocolContext.getBlockchain().observeBlockAdded(PandaPrinter.getInstance());
+    protocolContext
+        .getConsensusContext(MergeContext.class)
+        .addNewForkchoiceMessageListener(PandaPrinter.getInstance());
```

Comment:
it will be nice to unsubscribe the panda printer after it has done the job, or are you thinking of removing this code after the merge?

Reply:
I'll just remove it. Unsubscribing is a good idea, but then the PandaPrinter needs to keep a reference to the ConsensusContext so it can unsubscribe after finalization.